### PR TITLE
[MemProf] Expand optimization scope to internal linkage function

### PR DESCRIPTION
**Pull request by lifengxiang1025**
Now MemProf can't do IR annotation right in the local linkage function and global initial function __cxx_global_var_init. In llvm-profdata which convert raw memory profile to memory profile, it uses function name in dwarf to create GUID. But when llvm consumes memory profile,  it use `getIRPGOFuncName` or `getPGOFuncName` which returns local linkage function as `FileName;FunctionName` or `FileName:FunctionName`  to get function name and create GUID. So profile creator's GUID is not same as profile consumer.
So I think MemProf should be used with `unique-internal-linkage-names` and  don't use PGOFuncName.
__cxx_global_var_init is created later than where UniqueInternalLinkageNames works. So I add uniq suffix to __cxx_global_var_init additionally.

### CODE REVIEW

**Comment by lifengxiang1025** [comment]
@teresajohnson  Can you take a look and provide your insights? I'm learning how to write tests and will upload test cases soon.

**Comment by teresajohnson** [comment]
@snehasish can you take a look at the issue described here? Should we be doing something different in llvm-profdata?

**Comment by snehasish** [comment]
@lifengxiang1025 thanks for flagging this issue. I think it's best to not rely on unique-internal-linkage-name here. Instead we should extend the logic in RawMemProfReader.cpp to include "<filename>;" if the function is internal linkage as expected by IRPGOFuncName.  Can you try this approach?

https://github.com/llvm/llvm-project/blob/c0fe0719df457b0992606b0f2a235719a5bbfd54/llvm/lib/ProfileData/RawMemProfReader.cpp#L509-L510

**Comment by teresajohnson** [comment]
> @lifengxiang1025 thanks for flagging this issue. I think it's best to not rely on unique-internal-linkage-name here. Instead we should extend the logic in RawMemProfReader.cpp to include "filename;" if the function is internal linkage as expected by IRPGOFuncName. Can you try this approach?

I don't think we want to change the name in the frames themselves, as the names in the debug info when matching don't contain this prefix. Although I suppose we could modify the matcher to add the prefixes when matching frames too. I.e. here: https://github.com/llvm/llvm-project/blob/c0fe0719df457b0992606b0f2a235719a5bbfd54/llvm/lib/Transforms/Instrumentation/MemProfiler.cpp#L806

It sounds like the issue is that the names used to index the memprof records do not use the PGOFuncName scheme and therefore we never find any memprof records for an internal linkage function during matching. I.e. the GUID passed to InstrProfWriter::addMemProfRecord. Unfortunately, looking at RawMemProfReader.cpp it does look like that eventually gets populated out of the Function (GUID) field from same Frame setup code. So we'd either need to do some extra handling in that code so that the prefix is only used for the record entry, or change the matcher (the latter may be the easiest).

**Comment by lifengxiang1025** [comment]
> @lifengxiang1025 thanks for flagging this issue. I think it's best to not rely on unique-internal-linkage-name here. Instead we should extend the logic in RawMemProfReader.cpp to include "filename;" if the function is internal linkage as expected by IRPGOFuncName. Can you try this approach?
> 
> https://github.com/llvm/llvm-project/blob/c0fe0719df457b0992606b0f2a235719a5bbfd54/llvm/lib/ProfileData/RawMemProfReader.cpp#L509-L510

I don't find how to get linkage type in dwarf. It seems dwarf don't record linkage type. And the linkage type between thin compile and link step can change also.

**Comment by snehasish** [comment]
> change the matcher (the latter may be the easiest).

My response assumed that it was already using the IRPGOFuncName for matching. Yes, I think we should make the switch then.

> It seems dwarf don't record linkage type.

Yes, you're right. As an alternative can we use the symbol table and find `Bind = LOCAL` to add the prefix before hashing?

**Comment by lifengxiang1025** [comment]
> Yes, you're right. As an alternative can we use the symbol table and find Bind = LOCAL to add the prefix before hashing?

If we choose this method. I think we can't deal with the situation which one symbol is not local linkage type in thin compile, but will be converted to local linkage after thin backend by thinlto's internalization

**Comment by teresajohnson** [comment]
@snehasish and I chatted more about this offline. Using the dwarf info to figure out the right source name prefix during indexing is not straightforward. The source file name used as the prefix in the compiler for the IRPGOName is that of the TU. The source file of the line in the dwarf info may be a header. It could be possible to walk up the profiled inline frames to find a source file, but that is error prone in certain circumstances (e.g. if the profiled binary had thinlto cross module inlining, and if it isn't obvious which source file name is a non-included TU).

I think your change to the matching is almost strictly better than the complete non-matching we get currently for local functions. It would even work in many cases without uniq suffixes, but uniq suffixes will make that work even better. So let's go ahead with that change.

However, we'd like to request that you split the __cxx_global_var_init change into a separate patch, as it is somewhat orthogonal to the matching change, and in case there are unexpected issues with that change. Can you split that out and make the test case here for matching of a non __cxx_global_var_init function?

Thanks!

**Comment by lifengxiang1025** [comment]
> @snehasish and I chatted more about this offline. Using the dwarf info to figure out the right source name prefix during indexing is not straightforward. The source file name used as the prefix in the compiler for the IRPGOName is that of the TU. The source file of the line in the dwarf info may be a header. It could be possible to walk up the profiled inline frames to find a source file, but that is error prone in certain circumstances (e.g. if the profiled binary had thinlto cross module inlining, and if it isn't obvious which source file name is a non-included TU).
> 
> I think your change to the matching is almost strictly better than the complete non-matching we get currently for local functions. It would even work in many cases without uniq suffixes, but uniq suffixes will make that work even better. So let's go ahead with that change.
> 
> However, we'd like to request that you split the __cxx_global_var_init change into a separate patch, as it is somewhat orthogonal to the matching change, and in case there are unexpected issues with that change. Can you split that out and make the test case here for matching of a non __cxx_global_var_init function?
> 
> Thanks!

Thanks for your reply. I have removed the __cxx_global_var_init change. Please review again.

File: llvm/test/Transforms/PGOProfile/Inputs/update_memprof_inputs.sh [hunk 1]
```diff
@@ -127,3 +127,21 @@ ${CLANG} ${COMMON_FLAGS} -fmemory-profile -DUSE_MUSTTAIL=1 ${OUTDIR}/memprof_mis
 env MEMPROF_OPTIONS=log_path=stdout ${OUTDIR}/memprof_missing_leaf.exe > ${OUTDIR}/memprof_missing_leaf.memprofraw
 
 rm ${OUTDIR}/memprof_missing_leaf.cc
+
+cat > ${OUTDIR}/memprof_internal_linakge.cc << EOF
```
Comment (teresajohnson) [inline llvm/test/Transforms/PGOProfile/Inputs/update_memprof_inputs.sh:131]: s/linakge/linkage/ here and elsewhere in this file and the actual test.

File: llvm/lib/Transforms/Instrumentation/MemProfiler.cpp [hunk 2]
```diff
@@ -677,24 +677,10 @@ static void readMemprof(Module &M, Function &F,
                         const TargetLibraryInfo &TLI) {
   auto &Ctx = M.getContext();
 
-  auto FuncName = getIRPGOFuncName(F);
+  auto FuncName = F.getName();
```
Comment (teresajohnson) [inline llvm/lib/Transforms/Instrumentation/MemProfiler.cpp:680]: Can you add a comment summarizing why we specifically don't want to use getIRPGOFuncName here?

**Comment by Enna1** [comment]
also s/linakge/linkage/ in commit message and PR title.

**Review by teresajohnson — APPROVED** [review]

**Pull request merged**